File: problems/205/205.isomorphic-strings.rb
#
# @lc app=leetcode id=205 lang=ruby
#
# [205] Isomorphic Strings
#

# @lc code=start
# @param {String} s
# @param {String} t
# @return {Boolean}
def is_isomorphic(s, t)
  s_to_t = {}
  t_to_s = {}

  (0...s.length).each do |i|
    s_to_t[s[i]] = t[i] unless s_to_t[s[i]]
    t_to_s[t[i]] = s[i] unless t_to_s[t[i]]
    return false if s_to_t[s[i]] != t[i] || t_to_s[t[i]] != s[i]
  end

  true
end
# @lc code=end

File: problems/205/205.isomorphic-strings.py
#
# @lc app=leetcode id=205 lang=python3
#
# [205] Isomorphic Strings
#

# @lc code=start


class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        # ismorphic if chars in s can be replaced to get t
        # hashmap to store s->t, t->s char replacement
        # time O(N), space O(1)

        s_to_t, t_to_s = {}, {}
        for i in range(len(s)):
            if not s_to_t.get(s[i]):
                s_to_t[s[i]] = t[i]
            if not t_to_s.get(t[i]):
                t_to_s[t[i]] = s[i]
            if s_to_t[s[i]] != t[i] or t_to_s[t[i]] != s[i]:
                return False

        return True


# @lc code=end
